rs/sns/governance/src/proposal.rs

```diff
@@ -1058,77 +1060,87 @@ fn validate_and_render_upgrade_sns_controlled_canister(
     let mode = upgrade.mode_or_upgrade();
 
     // Inspect canister_id.
-    let mut canister_id = PrincipalId::new_user_test_id(0xDEADBEEF); // Initialize to garbage. This won't get used later.
-    match validate_required_field("canister_id", &upgrade.canister_id) {
+    let canister_id = match validate_required_field("canister_id", canister_id) {
         Err(err) => {
             defects.push(err);
+            None
         }
-        Ok(id) => {
-            canister_id = *id;
-        }
-    }
+        Ok(principal_id) => match CanisterId::try_from_principal_id(*principal_id) {
+            Ok(canister_id) => Some(canister_id),
+            Err(err) => {
+                let defect = format!(
+                    "UpgradeSnsControlledCanister.canister_id is invalid: {:?}",
+                    err
+                );
+                defects.push(defect);
+                None
+            }
+        },
+    };
 
     // Inspect wasm.
-    const RAW_WASM_HEADER: [u8; 4] = [0, 0x61, 0x73, 0x6d];
-    // see https://ic-interface-spec.netlify.app/#canister-module-format
-    const GZIPPED_WASM_HEADER: [u8; 3] = [0x1f, 0x8b, 0x08];
-
-    if new_canister_wasm.len() < 4
-        || new_canister_wasm[..4] != RAW_WASM_HEADER[..]
-            && new_canister_wasm[..3] != GZIPPED_WASM_HEADER[..]
-    {
-        defects.push("new_canister_wasm lacks the magic value in its header.".into());
-    }
-
-    if new_canister_wasm.len().saturating_add(
-        canister_upgrade_arg
-            .as_ref()
-            .map(|arg| arg.len())
-            .unwrap_or_default(),
-    ) >= MAX_INSTALL_CODE_WASM_AND_ARG_SIZE
-    {
-        defects.push(format!(
-            "the maximum canister WASM and argument size \
-             for UpgradeSnsControlledCanister is {} bytes.",
-            MAX_INSTALL_CODE_WASM_AND_ARG_SIZE
-        ));
-    }
+    let wasm_info = match Wasm::try_from(upgrade) {
+        Err(err) => {
+            defects.push(err);
+            None
+        }
+        Ok(wasm) => match wasm.validate(env, canister_upgrade_arg).await {
+            Err(new_defects) => {
+                defects.extend(new_defects.into_iter());
+                None
+            }
+            Ok(_) => Some(wasm.info()),
+        },
+    };
 
     // Generate final report.
     if !defects.is_empty() {
+        let tip = if upgrade.chunked_canister_wasm.is_some() {
+            format!(
+                "\nPlease make sure that both Governance ({}) and Root ({}) are controllers of \
```

Comment:
This comment made me think... if the canister is malicious, is root or governance needing to make a request to that store canister?

Reply:
> if the canister is malicious

I don't see how the store canister can exploit the SNS this way. The SNS makes only the following calls to the management canister that involve the store canister: 
1. `ic00.stored_chunks(record { canister_id = $STORE })`
    - Called the 1st time at proposal submission
    - Called the 2nd time at proposal execution (if adopted)
2. `ic00.install_chunked_code(record { store_canister = opt $STORE; ... })`

If, e.g., the store canister (or its controllers) decide to manipulate the WASM chunks specified in the proposal, then the hashes won't add up and the proposal would fail. 

Do you have a specific attack in mind?

Reply:
For context, only the controllers are allowed to list stored chunks and call `ic00.install_chunked_code` with a given store canister.

Reply:
@aterga 
> Do you have a specific attack in mind?

Open call contexts.  I think this will be fine when we have best effort messaging, which I think should be soon.  But in the meantime, I believe our policy is to not send requests out from governance directly.  That's part of why we have the root canister be the intermediary for dapp ownership.